### PR TITLE
[Feature] Add AllGather-KV sequence-parallel attention str…

### LINKED ISSUE #4690: [Feature]: Enable Communication-Compute Overlap in USP Scenarios
### 🚀 The feature, motivation and pitch

# Problem
In ulysses sp case, the all2all kernel occupy much time(15% in hunyuanimages with up4)
Image

# Propose
### 1.1 Overlap all2all with qkv project
Image

### 1.2 Option with overlap with head dim chunk
like session post-all2all

## 2 post-all2ll
Further divide head dim in each rank， head_chunk_1, head_chunk2 ....，and overlap head_chunk_1 all2all with head_chunk_2 compute

Image

**I think this is essentially a trade-off between maintaining a large fused kernel and achieving better communication overlap. The optimization benefit may vary across different models.**

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by AbelSara**
## Purpose
Introduce a new AllGather-KV sequence-parallel (SP) attention strategy for HunyuanImage3. Each rank holds 1/P of Q (pre_processor sequence-split); AllGather collects full K/V on every rank; a single local FA kernel computes Q_local x K_full. v1 is mutually exclusive with Ulysses/Ring and only valid for causal=False attention.

linked: #4690 
co-author: @Bounty-hunter 

#### performence tests
pytest -s -v tests/dfx/perf/scripts/run_diffusion_benchmark.py   --test-config-file tests/dfx/perf/tests/test_hunyuan_image_tp2_sp2.json

#### precision tests
pytest -s tests/e2e/accuracy/test_hunyuan_image3_pixel_accuracy.py

#### 
#### Ulysess Sp (TP=2 SP=2)
image

#### Context SP(TP=2 SP=2)
image

**speed up:1.24x**

#### Ulysess Sp (SP=2)
image

#### Context SP(SP=2)
image

**speed up:1.10x**

#### tp
image

#### ulysses sp
image

#### all gather sp
image

#### result
In terms of accuracy, SP has advantages over TP in both SSIM and PSNR.

### Generated Image
image

**Enviroment**
4 * NVIDIA RTX PRO 6000

**vLLM Version:**
0.24.0

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by AbelSara** [comment]
@qiaoyifan Could you please review my code when you have time?

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 1]
```diff
@@ -1165,11 +1166,12 @@ def __call__(
                 joint_text_query = query[:, :0, :, :]
                 joint_text_key, joint_text_value = self._reuse_prompt_kv(key, value, seq_len, bs, shard_image_size)
 
-        key = repeat_kv(key, repeat_num)
-        value = repeat_kv(value, repeat_num)
-        if self.sp_size > 1:
-            joint_text_key = repeat_kv(joint_text_key, repeat_num)
-            joint_text_value = repeat_kv(joint_text_value, repeat_num)
+        if not use_allgather_kv:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: just rm for other case?

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            if joint_strategy == "front":
+                k_full = torch.cat([joint_k, k_img_full], dim=1)
+                v_full = torch.cat([joint_v, v_img_full], dim=1)
+            else:
+                k_full = torch.cat([k_img_full, joint_k], dim=1)
+                v_full = torch.cat([v_img_full, joint_v], dim=1)
+        else:
+            k_full, v_full = k_img_full, v_img_full
+
+        k_full = _repeat_kv(k_full, kv_repeat_num)
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/attention/parallel/allgather_kv.py]: Actually, our backend support GQA, don't need to repeat.

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                f"Got joint_strategy={joint_strategy!r}."
+            )
+
+        img_start = self._sp_rank * img_seq_local
+        img_end = min(img_start + img_seq_local, img_seq_full)
+        local_spans_per_batch: list[list[tuple[int, int]]] = []
+
+        for spans in spans_per_batch:
+            local_spans: list[tuple[int, int]] = []
+            for start, end, mode in build_segments(spans, query_offset=0, query_len=joint_len):
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/attention/parallel/allgather_kv.py]: I think this is uncompatible with build_segments，it's input is `global index spans`, `real query_offset`, `query_len`. please check the accuracy in bitwise.

File: vllm_omni/diffusion/distributed/parallel_state.py [hunk 4]
```diff
@@ -835,9 +848,20 @@ def initialize_model_parallel(
...
+    if allgather_degree > 1:
+        sp_ulysses_degree_for_pg = allgather_degree
+        sp_ring_degree_for_pg = 1
+    else:
+        sp_ulysses_degree_for_pg = ulysses_degree
+        sp_ring_degree_for_pg = ring_degree
     ulysses_pg, ring_pg = set_seq_parallel_pg(
-        sp_ulysses_degree=ulysses_degree,
-        sp_ring_degree=ring_degree,
+        sp_ulysses_degree=sp_ulysses_degree_for_pg,
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/distributed/parallel_state.py]: `allgather`  at the same level of abstraction with `ulysses` and `ring`,  should add new group. Additionally, please consider whether context parallelism is compatible with enabling both Ulysses or Ring simultaneously. For the first stage, I think it is acceptable to leave this compatibility out of scope.

**Review by RuixiangMa — COMMENTED** [review]
@AbelSara  Missing test coverage in `allgather_kv.py`, current tests only cover joint_strategy="front" with a dense mask. Please add tests for:
  - joint_strategy="back" mask slicing 
  - full_attn_spans path (non-dense mask)
  - kv_repeat_num > 1 (GQA) round-trip

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py (hunk 2 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/parallel/allgather_kv.py]: fixed

**Comment by AbelSara** [comment]
> @AbelSara Missing test coverage in `allgather_kv.py`, current tests only cover joint_strategy="front" with a dense mask. Please add tests for:
> 
> * joint_strategy="back" mask slicing
> * full_attn_spans path (non-dense mask)
> * kv_repeat_num > 1 (GQA) round-trip

Relevant tests have been added.

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py (hunk 3 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/parallel/allgather_kv.py]: fixed

File: vllm_omni/diffusion/distributed/parallel_state.py (hunk 4 above)
Comment (AbelSara) [reply vllm_omni/diffusion/distributed/parallel_state.py]: fixed

**Comment by AbelSara** [comment]
This PR is ready for review @Bounty-hunter @RuixiangMa @hsliuustc0106

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 1 above)
Comment (AbelSara) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: linked to #3939

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py (hunk 2 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/parallel/allgather_kv.py]: In the SDPA path, repeating K/V to match the Q head dimension before computation is faster.

File: vllm_omni/diffusion/distributed/parallel_state.py [hunk 5]
```diff
@@ -586,6 +593,31 @@ def set_seq_parallel_pg(
...
+    if sp_allgather_degree > 1:
+        if sp_ulysses_degree > 1 or sp_ring_degree > 1:
+            raise ValueError("AllGather-KV is mutually exclusive with Ulysses and Ring")
+        sp_size = sp_allgather_degree
+        if sp_group_ranks is None:
+            sp_group_ranks = [list(range(offset, offset + sp_size)) for offset in range(0, world_size, sp_size)]
+        if len(sp_group_ranks) * sp_size != world_size or any(len(ranks) != sp_size for ranks in sp_group_ranks):
+            raise ValueError(f"Invalid sp_group_ranks: expected {world_size // sp_size} groups of size {sp_size}.")
+
+        ulyssess_pg = ring_pg = allgather_pg = None
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/distributed/parallel_state.py]: NIT: typo ulyssess_pg (triple 's'),  non-functional, suggest renaming to ulysses_pg for consistency in a follow-up cleanup

**Comment by RuixiangMa** [comment]
LGTM

File: vllm_omni/diffusion/distributed/parallel_state.py (hunk 5 above)
Comment (AbelSara) [reply vllm_omni/diffusion/distributed/parallel_state.py]: fixed,thx

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        joint_strategy = "front"
+        if attn_metadata is not None:
+            joint_q = attn_metadata.joint_query
+            joint_k = attn_metadata.joint_key
+            joint_v = attn_metadata.joint_value
+            joint_strategy = attn_metadata.joint_strategy or "front"
+        if joint_strategy not in {"front", "rear"}:
+            raise ValueError(f"Unsupported joint_strategy: {joint_strategy!r}")
+
+        k_img_full = self._sp_group.all_gather(key, dim=1, group=self._allgather_group)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/parallel/allgather_kv.py:63]: I think this all_gather leans on every rank handing in an identically-shaped `key` — `all_gather_into_tensor` just splits the output evenly by world size. If the image sequence isn't divisible by `allgather_degree`, the shards differ and this either mismatches or hangs. `img_start = self._sp_rank * img_seq_local` further down assumes the same uniform shard too 🤔 Is even divisibility guaranteed upstream, or should we assert it here before gathering?

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 7]
```diff
@@ -191,6 +192,17 @@ def forward_cuda(
                 attn_func=flash_attn_func,
             )
 
+            if attn_metadata.query_ranges is not None:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: I noticed this `query_ranges` → `mapped_piecewise_attn` branch only exists in the flash backends. SDPA and cuDNN never look at `query_ranges`, so if AllGather-KV runs on one of those, the globally-non-contiguous local query gets treated as contiguous and the numerics silently diverge from the non-SP path. I'd either restrict AllGather-KV to the flash backends explicitly, or teach SDPA/cuDNN to honor `query_ranges` too.

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py (hunk 6 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/parallel/allgather_kv.py:63]: Its SP plan uses auto_pad=True, the image sequence is zero-padded to a multiple of the full sequence-parallel world size before chunk(world_size), so every rank, and consequently every rank in an all-gather subgroup, receives the same local sequence length.

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 7 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: sdpa support 4d mask

**Comment by Bounty-hunter** [comment]
Can you enable all_gather and run `test_hunyuan_image3_pixel_accuracy.py` test ?

File: vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py [hunk 8]
```diff
@@ -56,6 +68,26 @@ def build_segments(full_attn_spans, query_offset, query_len):
     return segs
 
 
+def build_mapped_segments(full_attn_spans, query_offset, query_len):
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py]: We need to unifiled `segments` and `mapped_segments`. The origin design assume that the segment_i only attend to the tokens before its last token. Now we introduce sequence parallelism together with full attention, the query may need to attend to KV tokens after the segment boundary, so we need to extend the implementation accordingly. I think the new implementation is compatible with the previous one, so we can unify them into a single solution.

File: vllm_omni/diffusion/attention/parallel/allgather_kv.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        else:
+            q_full_len = joint_len + img_seq_full
+            if mask.shape[-2] != q_full_len:
+                raise ValueError(
+                    "AllGather-KV SP received an attention mask with incompatible Q length: "
+                    f"mask_q={mask.shape[-2]}, expected local_q={q_local_len} or full_q={q_full_len} "
+                    f"(joint_len={joint_len}, img_seq_local={img_seq_local}, img_seq_full={img_seq_full})."
+                )
+
+            if joint_len > 0 and joint_strategy == "front":
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/attention/parallel/allgather_kv.py]: Simplify：since we have ranges, we can just
```
parts = [
    mask[..., r.global_start:r.global_start + (r.local_end - r.local_start), :]
    for r in ranges
]
local_mask = torch.cat(parts, dim=-2)
```

Comment (AbelSara) [reply vllm_omni/diffusion/attention/parallel/allgather_kv.py]: fixed

File: vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py (hunk 8 above)
Comment (AbelSara) [reply vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py]: fixed

**Comment by AbelSara** [comment]
> Can you enable all_gather and run `test_hunyuan_image3_pixel_accuracy.py` test ?

Thanks! I’ve added the corresponding tests and updated the results in the PR description.

**Comment by hsliuustc0106** [comment]
@Bounty-hunter does this work in NPU and other hardware as well?

**Comment by Bounty-hunter** [comment]
> @Bounty-hunter does this work in NPU and other hardware as well?

Yes, this is hardware-independent logic.

**Comment by hsliuustc0106** [comment]
> > @Bounty-hunter does this work in NPU and other hardware as well?
> 
> Yes, this is hardware-independent logic.

any test results in NPU?

**Comment by Bounty-hunter** [comment]
I tested on L20X, and the performance gain from all-gather over USP is not as significant as what you observed on RTX Pro 6000. Instead, I’m seeing a slight regression. I’ll do some profiling to understand the reason.
Could you also profile it on RTX Pro 6000 to check whether there are any extra operations introduced that might contribute to the regression, such as synchronization overhead? @AbelSara

**Comment by Bounty-hunter** [comment]
It 4.23% improve in l20x with config: fp8，4card:
|  | USP4 (Ulysses=4) | AG4 (AllGather=4) | AG4 vs USP4 |
|---|---|---|---|
| **throughput_qps** | 0.4209 | 0.4387 | **+4.23%** |


I notice that, the attention duration from 1.71ms  to 469 us

**ulysses**:
image

**allgather**:
image

**Review by Bounty-hunter — APPROVED** [review]
LGTM

File: vllm_omni/diffusion/attention/backends/sdpa.py [hunk 10]
```diff
@@ -103,7 +103,24 @@ def _forward_impl(
         if attn_metadata:
             attention_mask = _maybe_reshape_attn_mask(query, key, attn_metadata.attn_mask, mask_mode=mask_mode)
```
Comment (Gaohan123) [inline vllm_omni/diffusion/attention/backends/sdpa.py:105]: `AllGatherKVParallelAttention` can emit non-contiguous `query_ranges` together with `full_attn_spans`, but this SDPA path only consumes `attn_mask` and ignores both fields. The Flash backends route this case through `piecewise_attn(..., query_ranges=...)`, while SDPA (and similarly cuDNN) will treat the concatenated local Q as a contiguous sequence.

That means an AllGather-KV run using the non-dense/mixed causal-full attention path can silently compute different attention semantics when it selects SDPA/cuDNN instead of Flash. Could we either restrict AllGather-KV + `full_attn_spans` to the Flash backends, or add equivalent piecewise/query-range handling for SDPA and cuDNN (with coverage for each backend)?

Comment (AbelSara) [reply vllm_omni/diffusion/attention/backends/sdpa.py:105]: fixed

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
